src/v/raft/consensus.cc

```diff
@@ -2153,18 +2153,46 @@ consensus::install_snapshot(install_snapshot_request&& r) {
 ss::future<> consensus::hydrate_snapshot() {
     // Read snapshot, reset state machine using snapshot contents (and load
     // snapshot’s cluster configuration) (§7.8)
-    return _snapshot_mgr.open_snapshot().then(
-      [this](std::optional<storage::snapshot_reader> reader) {
-          // no snapshot do nothing
-          if (!reader) {
-              return ss::now();
-          }
-          return ss::do_with(
-            std::move(*reader), [this](storage::snapshot_reader& reader) {
-                return do_hydrate_snapshot(reader).finally(
-                  [&reader] { return reader.close(); });
-            });
-      });
+    auto metadata = co_await read_snapshot_metadata();
+    if (!metadata.has_value()) {
+        co_return;
+    }
+    load_from_metadata(metadata.value());
+    co_await _configuration_manager.add(
+      _last_snapshot_index, std::move(metadata->latest_configuration));
+    _probe->configuration_update();
+    auto delta = delta_for_truncation(metadata.value());
+    if (delta.has_value()) {
+        co_await _offset_translator.prefix_truncate_reset(
+            _last_snapshot_index, *delta);
+        co_await truncate_to_latest_snapshot();
+    }
+    _snapshot_size = co_await _snapshot_mgr.get_snapshot_size();
+}
+
+std::optional<model::offset_delta>
+consensus::delta_for_truncation(const snapshot_metadata& metadata) {
+    if (
+      _keep_snapshotted_log
+      && _log->offsets().dirty_offset >= _last_snapshot_index) {
+        // skip prefix truncating if we want to preserve the log (e.g. for the
+        // controller partition), but only if there is no gap between old end
+        // offset and new start offset, otherwise we must still advance the log
+        // start offset by prefix-truncating.
+        return std::nullopt;
+    }
+    auto delta = metadata.log_start_delta;
+    if (delta < offset_translator_delta(0)) {
+        delta = offset_translator_delta(
+          _configuration_manager.offset_delta(_last_snapshot_index));
+        vlog(
+          _ctxlog.warn,
+          "received snapshot without delta field in metadata, "
+          "falling back to delta obtained from configuration "
+          "manager: {}",
+          delta);
+    }
```

Comment:
I think you're referring to
>raft: refactor snapshot hydration

Is that right?

What's changed is that previously we would unconditionally truncate the offset translator state. Now this will only happen if we're also truncating the log.